### ISSUE TITLE
documentation: use of element.exists and element.count

### ISSUE THREAD

**Issue by fmancardi** [DevExpress/testcafe#6273]
In the guides I've found this about element.exists and element.count
[Selector Timeout](https://testcafe.io/documentation/402829/guides/basic-guides/select-page-elements#selector-timeout)

_When a selector is executed, TestCafe waits for the target node to appear in the DOM until the selector timeout expires.
...
Selector timeouts have no effect on Selector.exists and Selector.count properties. **These properties are evaluated immediately regardless of a timeout.** To apply a timeout to exists and count assertions, pass the timeout to the assertion method (expect.ok, expect.eql, etc.)._

In [Selector.exists Property](https://testcafe.io/documentation/402754/reference/test-api/selector/exists)
There is this code  
```

import { Selector } from 'testcafe';

fixture `Example page`
    .page `http://devexpress.github.io/testcafe/example/`;

test('My test', async t => {
    const submitButtonExists = Selector('#submit-button').exists;

    await t.expect(submitButtonExists).ok();
});
```
**QUESTION/DOUBT 1:**  
but it's not clear the intent of initializing const submitButtonExists  or if exists is evaluated twice,  
once IMMEDIATELY when initializing submitButtonExists and the other in the expect() (here with the WAIT).   

In **Check if an Element is Available** section the code example is

```
import { Selector } from 'testcafe';

fixture `My fixture`
    .page `https://devexpress.github.io/testcafe/example/`;

test('My test', async t => {
    const element = Selector('#developer-name');

    if(await element.exists && await element.visible)
        await t.typeText(element, 'Peter Parker');

    // ...
});
```

**QUESTION / DOUBT 2** 
if exits does not uses the smart await mechanism, then this it probably sometimes will fail.  
In addition (I've already asked this in another issue) it will not be enough the following check? 
**How do you make exists wait without using an expect?**

```
    if(await element.visible)
        await t.typeText(element, 'Peter Parker');

    // ...

```
**QUESTION / DOUBT 3** 
it's possible for an element to be visible but does not exist?  

Probably it's my fault and I'm misunderstanding this documentation, but IMHO some details about how exists and count work regarding immediately or not execution work will be very useful in the 
[Selector.exists Property](https://testcafe.io/documentation/402754/reference/test-api/selector/exists)
section cross-linked to the other sections where there are references to exists.

**Comment by AlexKamaev** [comment]
Hi

> Q1 - After the STEP 2, what is the mechanisim used by testcafe, that will make it consider that the target node is not present in the anymore in the DOM ?

TestCafe does not cache results of the Selector execution. So, when you execute your Selector for the second time, TestCafe tries to find the target element again. Therefore, if the element is missing for some reason, TestCafe shows the error message.

> Q2 It's possible that after page reload I see an element in the screen, but test cafe considers that the element does not exist?

In some cases, TestCafe can consider the element as `not visible` (`el.visible` returns `false`) even if you see the element on your screen. However, TestCafe should consider it as `existing` (`el.exists` returns `true`) anyway.

> Q3 it's possible this situation
```
if (await elem.visible) {
   await t.expect(elem.exists).ok();   **// and that this expect will fail?**
}
```
In general, this should not happen. It can only happen if the element is deleted on the client side after the `elem.visible` check is complete.

**Comment by fmancardi** [comment]
Thanks for your help, but it seems you have copied here questions and answer from #6237.

**Comment by AlexKamaev** [comment]
Hi

> QUESTION/DOUBT 1:
> but it's not clear the intent of initializing const submitButtonExists or if exists is evaluated twice,
> once IMMEDIATELY when initializing submitButtonExists and the other in the expect() (here with the WAIT).

`exists` in this case is not evaluated twice. It starts evaluating on this line:
```const submitButtonExists = Selector('#submit-button').exists;```
and finishes here:
```await t.expect(submitButtonExists).ok();```

The `exists` property of the Selector is a Promise. So, on the first line you start to execute it, and in your `expect` statement you wait until this promise is resolved. But it is the same promise, and it's executed only once.

> QUESTION / DOUBT 2
> if exits does not uses the smart await mechanism, then this it probably sometimes will fail.
> In addition (I've already asked this in another issue) it will not be enough the following check?
> How do you make exists wait without using an expect?

Some time ago we improved the `visible` property. Now, `visible` can wait if an element does not exist but can appear in the future. Please see the following example:


```HTML
<script>
        const deferredElement = document.createElement('div');

        deferredElement.innerText = 'click me';

        setTimeout(function () { document.body.appendChild(deferredElement) }, 5000);
    </script>
```
```JS
test.only('Question/Doubt 2', async t => {
    const element = Selector('div');

    if (await element.visible) {
        await t.click(element);
    }
});

```
In this example TestCafe will wait for the element to appear.

> QUESTION / DOUBT 3
> it's possible for an element to be visible but does not exist?

In general, the answer is no. But there is a specific case - if the element is created during the selector timeout. For example, in my Q2 example, the element is visible but does not exist.
